File: back/runtime/agentloop/emittedtarget.go
// emittedtarget.go — EL00: the DECLARED target of the EMITTED app (ADR 0040), as a single
// Go-source truth. EL00 is ADR-only — no package logic, just the engraving of the target the
// BesoinGraph elicits against and the app-builder emits toward — exactly as FN02/ADR 0036
// engraved the functional mandate. This file IS that engraving in code form, so the parity
// test (emittedtarget_parity_test.go) can pin it ≡ the arch-fitness.json config.
//
// THE FRONTIER constructrice ≠ construite (ADR 0040 Décision 1/2):
//   - La CONSTRUCTRICE — AIDOS itself: Go (Kernel/Runtime/Archive + the Operation-DSL
//     interpreter) + the Next Workbench + the Postgres truth-store. A governable OS, under
//     the wall. NEVER rewritten in TS; it NEVER emits Go for the user app.
//   - La CONSTRUITE — the app AIDOS emits for the user (the gen/ tree): Hono (front+back),
//     pure-functional TypeScript, with its OWN MCP + Skills; datastore = Postgres dialect
//     (ADR 0006 UNCHANGED) via a TS client; Operation-DSL executed by a CALLBACK to a Go
//     interpreter service (ADR 0040 Décision 7 — never a TS re-emission of a governed truth).
//
// DETERMINISM-FIRST + THE WALL (CLAUDE.md §2/§6/§8). EL00 writes NO truth: it records a
// decision (an ADR + provenance), above-the-line. The emitted app stays governed by
// idée → miroir → /goal; the Hono/TS pivot creates NO truth-write door. The emitted code
// stays a DETERMINISTIC emission (same Kernel → byte-identical bundle, "compile" = tsc); FN02
// purity IS determinism applied to the output, now on TS. The three emitted invariants are
// LANGUAGE-NEUTRAL: pivoting Go→Hono/TS adds none and forks none (Décision 3) — this file
// carries EmittedInvariantCodes() verbatim, no second list. Widening/softening the frontier
// is a TRUTH change (idée → miroir → /goal), never a silent edit — pinned by the parity test.
package agentloop

// Constructrice is the declared, frozen description of AIDOS itself — the governable OS that
// builds apps. It stays Go and is never rewritten in TS (ADR 0040 Décision 1/B).
type Constructrice struct {
	// Language of AIDOS the constructrice — stays "go" (Kernel/Runtime/Archive + interpreter).
	Language string `json:"language"`
	// RewrittenInTS — AIDOS is NEVER rewritten in TypeScript (ADR 0040 Décision 1, Option B
	// rejected). Always false.
	RewrittenInTS bool `json:"rewritten_in_ts"`
	// EmitsGoForUserApp — AIDOS NEVER emits Go for the user app (the frontier). Always false.
	EmitsGoForUserApp bool `json:"emits_go_for_user_app"`
}

// Construite is the declared, frozen target of the EMITTED user app (ADR 0040 Décision 1–7):
// Hono front+back, pure-functional TypeScript, own MCP+Skills, Postgres-dialect datastore via
// a TS client, Operation-DSL via a Go interpreter-service callback.
type Construite struct {
	// Backend / Frontend framework of the emitted app — both Hono (ADR 0040 Décision 1).
	Backend  string `json:"backend"`
	Frontend string `json:"frontend"`
	// Language — pure-functional TypeScript (ADR 0040 Décision 1/3).
	Language string `json:"language"`
	// DatastoreDialect — stays "postgres" (ADR 0006 UNCHANGED; only the access driver moves
	// from pgx to a TS client — ADR 0040 Décision 2).
	DatastoreDialect string `json:"datastore_dialect"`
	// OwnMCP / OwnSkills — the emitted app gets its OWN MCP + Skills (ADR 0040 Décision 4).
	OwnMCP    bool `json:"own_mcp"`
	OwnSkills bool `json:"own_skills"`
	// OperationInterpreter — how the emitted app runs the Operation-DSL. TRENCHED (ADR 0040
	// Décision 7 / OQ-0040-interpréteur) = "go-service-callback": the interpreter stays Go,
	// governable, non-duplicated, exposed as a service the Hono/TS handlers call via a typed
	// client. NEVER a TS re-emission of a governed truth.
	OperationInterpreter string `json:"operation_interpreter"`
}

// EmittedTarget is the single DECLARED truth of EL00: the constructrice≠construite frontier,
// the FN02 invariants applied to the emitted (Hono/TS) code, and the above-the-wall posture.
// The arch-fitness.json `emitted_target` block is the SAME record (parity test pins it).
type EmittedTarget struct {
	// ADR that engraves this target — "0040".
	ADR string `json:"adr"`
	// Constructrice / Construite — the two sides of the trenched frontier.
	Constructrice Constructrice `json:"constructrice"`
	Construite    Construite    `json:"construite"`
	// FunctionalInvariants — the FN02/ADR 0036 emitted invariants, LANGUAGE-NEUTRAL: the exact
	// list EmittedInvariantCodes() returns (no fork on the Go→TS pivot, ADR 0040 Décision 3).
	FunctionalInvariants []string `json:"functional_invariants"`
	// WritesTruth — EL00 is ADR-only, above-the-wall: it writes NO truth (CLAUDE.md §2). false.
	WritesTruth bool `json:"writes_truth"`
}

// EmittedTargetEL00 returns the declared EL00 target — a pure value, built and returned by
// value (it obeys the very EMITTED_NO_GLOBAL_MUTABLE invariant it carries). It is the single
// source the parity test compares against arch-fitness.json's `emitted_target` block.
func EmittedTargetEL00() EmittedTarget {
	invariants := EmittedInvariantCodes()
	names := make([]string, len(invariants))
	for i, c := range invariants {
		names[i] = string(c)
	}
	return EmittedTarget{
		ADR: "0040",
		Constructrice: Constructrice{
			Language:          "go",
			RewrittenInTS:     false,
			EmitsGoForUserApp: false,
		},
		Construite: Construite{
			Backend:              "hono",
			Frontend:             "hono",
			Language:             "typescript",
			DatastoreDialect:     "postgres",
			OwnMCP:               true,
			OwnSkills:            true,
			OperationInterpreter: "go-service-callback",
		},
		FunctionalInvariants: names,
		WritesTruth:          false,
	}
}

File: front/web/lib/emitted-target.ts
// emitted-target.ts — EL00: the pure TS twin of back/runtime/agentloop.EmittedTargetEL00().
//
// EL00 (ADR 0040) engraves the TARGET of the EMITTED app: Hono front+back, pure-functional
// TypeScript, its own MCP + Skills, datastore = Postgres dialect via a TS client, Operation-DSL
// via a Go interpreter-service callback. It draws the constructrice≠construite frontier: AIDOS
// stays governable Go (never rewritten in TS, never emits Go for the user app).
//
// This module is the byte-identical TS twin of the Go declared truth, so the Workbench panel can
// run the SAME parity verdict the Go test (config_test / emittedtarget_parity_test) runs — a pure
// function, no clock/rng/IO, no LLM. The wall (CLAUDE.md §2): EL00 writes NO truth; this twin only
// reads/renders the declared target and computes a deterministic parity verdict.

/** AIDOS la constructrice — stays Go, never rewritten in TS (ADR 0040 Décision 1/B). */
export interface Constructrice {
	language: string;
	rewritten_in_ts: boolean;
	emits_go_for_user_app: boolean;
}

/** The emitted user app la construite — Hono/TS target (ADR 0040 Décision 1–7). */
export interface Construite {
	backend: string;
	frontend: string;
	language: string;
	datastore_dialect: string;
	own_mcp: boolean;
	own_skills: boolean;
	operation_interpreter: string;
}

/** The single declared EL00 target — the same shape as the Go EmittedTarget record. */
export interface EmittedTarget {
	adr: string;
	constructrice: Constructrice;
	construite: Construite;
	functional_invariants: string[];
	writes_truth: boolean;
}

/** The FN02/ADR 0036 emitted invariants — LANGUAGE-NEUTRAL, the exact EmittedInvariantCodes()
 *  list (no fork on the Go→Hono/TS pivot, ADR 0040 Décision 3). Frozen, returned by value. */
export const EMITTED_INVARIANTS: readonly string[] = [
	"EMITTED_NO_GLOBAL_MUTABLE",
	"EMITTED_FUNCTION_PURE",
	"EMITTED_CALL_GRAPH_ACYCLIC",
] as const;

/** emittedTargetEL00 — the pure TS twin of the Go EmittedTargetEL00(). Same input (none) → same
 *  output, every call (the determinism-first reproducibility property). */
export function emittedTargetEL00(): EmittedTarget {
	return {
		adr: "0040",
		constructrice: {
			language: "go",
			rewritten_in_ts: false,
			emits_go_for_user_app: false,
		},
		construite: {
			backend: "hono",
			frontend: "hono",
			language: "typescript",
			datastore_dialect: "postgres",
			own_mcp: true,
			own_skills: true,
			operation_interpreter: "go-service-callback",
		},
		functional_invariants: [...EMITTED_INVARIANTS],
		writes_truth: false,
	};
}

/** A single parity check the panel renders. */
export interface ParityCheck {
	label: string;
	declared: string;
	enforced: string;
	ok: boolean;
}

/** The deterministic parity verdict: the declared EL00 target ≡ the enforced config block.
 *  The Workbench feeds it the `emitted_target` block read from arch-fitness.json (the config
 *  FN04/S84 enforces). Same as the Go reflect.DeepEqual parity test, surfaced on screen. */
export interface ParityVerdict {
	checks: ParityCheck[];
	ok: boolean;
}

function check(label: string, declared: string, enforced: string): ParityCheck {
	return { label, declared, enforced, ok: declared === enforced };
}

/** verifyParity — compares the declared target against an enforced config record, field by
 *  field, returning a per-field verdict. Pure: same inputs → same verdict, no LLM. */
export function verifyParity(
	declared: EmittedTarget,
	enforced: EmittedTarget,
): ParityVerdict {
	const checks: ParityCheck[] = [
		check("adr", declared.adr, enforced.adr),
		check(
			"constructrice.language",
			declared.constructrice.language,
			enforced.constructrice.language,
		),
		check(
			"constructrice.rewritten_in_ts",
			String(declared.constructrice.rewritten_in_ts),
			String(enforced.constructrice.rewritten_in_ts),
		),
		check(
			"constructrice.emits_go_for_user_app",
			String(declared.constructrice.emits_go_for_user_app),
			String(enforced.constructrice.emits_go_for_user_app),
		),
		check(
			"construite.backend",
			declared.construite.backend,
			enforced.construite.backend,
		),
		check(
			"construite.frontend",
			declared.construite.frontend,
			enforced.construite.frontend,
		),
		check(
			"construite.language",
			declared.construite.language,
			enforced.construite.language,
		),
		check(
			"construite.datastore_dialect",
			declared.construite.datastore_dialect,
			enforced.construite.datastore_dialect,
		),
		check(
			"construite.own_mcp",
			String(declared.construite.own_mcp),
			String(enforced.construite.own_mcp),
		),
		check(
			"construite.own_skills",
			String(declared.construite.own_skills),
			String(enforced.construite.own_skills),
		),
		check(
			"construite.operation_interpreter",
			declared.construite.operation_interpreter,
			enforced.construite.operation_interpreter,
		),
		check(
			"functional_invariants",
			declared.functional_invariants.join(","),
			enforced.functional_invariants.join(","),
		),
		check(
			"writes_truth",
			String(declared.writes_truth),
			String(enforced.writes_truth),
		),
	];
	return { checks, ok: checks.every((c) => c.ok) };
}
